src/AspNetCore/Src/applicationinfo.cpp

```diff
@@ -533,3 +522,101 @@ APPLICATION_INFO::FindNativeAssemblyFromHostfxr(
     }
     return hr;
 }
+
+VOID
+APPLICATION_INFO::RecycleApplication()
+{
+    APPLICATION* pApplication = NULL;
+    HANDLE       hThread = INVALID_HANDLE_VALUE;
+    BOOL         fLockAcquired = FALSE;
+
+    if (m_pApplication != NULL)
+    {
+        AcquireSRWLockExclusive(&m_srwLock);
+        fLockAcquired = TRUE;
+        if (m_pApplication != NULL)
+        {
+            pApplication = m_pApplication;
+            if (pApplication->QueryConfig()->QueryHostingModel() == HOSTING_OUT_PROCESS)
+            {
+                //
+                // For inprocess, need to set m_pApplication to NULL first to
+                // avoid mapping new request to the recycled application.
+                // Outofprocess application instance will be created for new request
+                // For inprocess, as recycle will lead to shutdown later, leave m_pApplication
+                // to not block incoming requests till worker process shutdown
+                //
+                m_pApplication = NULL;
```

Comment:
I think there is a potential race here with setting m_pApplication to NULL and getting a new request. Here is the scenario:

App_offline is dropped, and we immediately receive a request. The request calls EnsureApplicationCreated() and that succeeds. RecycleApplication is then called, causing the application to be null. We then call QueryApplication and that application is NULL. We have a debug assert there assuming the application, which is incorrect. 

I think we need to consider a few more race conditions here too.

Reply:
yes. there is possibility of NULL ref. I changed the function of get application to reference the application first.